feat : notifications screen

BraCR10 opened the pull request:
#  Notifications Screen Implementation

Closes #91

## Overview

This pull request introduces a notifications system that uses the previous temporary SnackBar-based approach with a persistent notification center. The implementation provides users with a complete history of their  activities produced by `lib/features/order/notfiers/abstract_mostro_notifier.dart`

## Problem Statement

The previous notification system relied on temporary SnackBar messages that disappeared after a few seconds, providing no historical context or detailed information about trade events. Users had no way to review past notifications or access detailed information about specific actions that occurred during their trading sessions.

## Solution Architecture

The new notification system is built around a persistent storage model using Sembast database integration with a  UI layer. The architecture follows a modular approach with clear separation of concerns between data persistence, state management, and user interface components.

###  Data Layer

The foundation of the new system is built on `lib/data/models/notification.dart`, which defines the notification model. This model works in conjunction with `lib/data/models/enums/notification_type.dart` that categorizes notifications into seven distinct types: orderUpdate, payment, dispute, cancellation, tradeUpdate, message, and system notifications.

The persistence layer is handled through `lib/data/repositories/notifications_history_repository.dart`, which implements a repository pattern with a clean interface abstraction. The repository provides full CRUD operations, bulk management features, and filtering options. The implementation uses `BaseStorage` as the  storage mechanism, ensuring reliable persistence across app restarts.

### State Management Integration

The notification system integrates seamlessly with the existing Riverpod-based state management architecture. The central component `lib/features/notifications/notifiers/notifications_notifier.dart` replaces the previous  notification approach with unified logic that handles all notification lifecycle events. This notifier works with `lib/features/notifications/notifiers/notifications_state.dart` to maintain immutable state and `lib/features/notifications/providers/notifications_provider.dart` for proper dependency injection.

### Message Localization and Mapping

In this implementation is the comprehensive message mapping system built around `lib/features/notifications/utils/notification_message_mapper.dart`. This utility maps each Mostro action to appropriate localized title and message keys, ensuring that every notification type has specific, contextually relevant messaging. The system covers over 25 different action types to provide better info of each notification. 

The notification details are handled through `lib/features/notifications/widgets/notification_details.dart`, where each action type receives customized detail presentation. This includes smart formatting for hash truncation, amount display, timeout conversion, and localized labeling. ***So keep this in mind when you need to provide more details about an action, as it should be presented in the best possible way and formatted well in that widget.***
<img width="320" height="678" alt="image" src="https://github.com/user-attachments/assets/e0ad67ac-8a34-45b9-9536-5a465bb2a9ea" />




## User Interface Implementation

The notification screen itself is implemented in `lib/features/notifications/screens/notifications_screen.dart`, providing users with a comprehensive view of their notification history. 

The integration with the existing app architecture required enhancements to `lib/shared/widgets/mostro_app_bar.dart`, making it more modular and flexible. The app bar now supports notification bell integration through configurable properties while maintaining backward compatibility with existing usage patterns.

A new interactive bell widget `lib/shared/widgets/notification_history_bell_widget.dart` provides live notification badges with unread count indicators. This widget automatically integrates with the enhanced app bar when notification functionality is enabled, providing users with immediate visual feedback about new notifications.

##  System Migration

The implementation involved removing the legacy notification system components, specifically `lib/shared/notifiers/notification_notifier.dart` and `lib/shared/providers/notification_notifier_provider.dart`. These components implemented the old SnackBar-based approach that provided only temporary messaging without persistence or detailed information.

Existing notification triggers continue to work seamlessly, but now feed into the new persistent notification system instead of generating temporary messages.

## Internationalization Expansion

This implementation includes a significant expansion of the internationalization system, adding over 90 new localization keys across three languages. The localization covers notification titles, detailed messages, UI labels, action buttons, and contextual information display.

English and Spanish translations have been personally verified for accuracy and consistency. ***Italian translations were generated using AI-assisted translation and require review*** The localization system is designed to be easily extensible for additional languages while maintaining consistency across all notification types.

## Integration Points

The notification system integrates with the existing order management system through enhancements to `lib/features/order/notfiers/abstract_mostro_notifier.dart`. 

The routing system in `lib/core/app_routes.dart` has been updated to include the new notifications screen, maintaining the existing app structure.

## Technical Considerations

One important architectural note concerns the existing `lib/notifications/notification_service.dart`, which handles OS-level push notifications through the system notification bar. This is separate from the new in-app notification system implemented in `lib/features/notifications/`. The naming similarity could potentially cause confusion for developers navigating the codebase, and consideration should be given to reorganizing these components for clarity.

## Sumary

26 new files added, 13 existing files modified, and 2 legacy files removed. Code includes comprehensive UI components, business logic, and localization content across multiple languages.

The modular architecture ensures maintainable code with clear separation between data persistence, business logic, and user interface components. 

I welcome feedback on the architectural decisions, user experience design, Italian translations, performance optimizations, and any additional functionality that would enhance the notification system.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - Added Notifications screen with history, pull-to-refresh, unread indicator, and per-item actions (mark as read, delete).
  - Introduced notification bell with unread badge and navigation to the Notifications screen.
  - Persisted notifications with real-time updates and unread count.
  - Localized notification titles/messages and details in English, Spanish, and Italian.
  - Time-ago timestamps and new status color for info.
- Refactor
  - Replaced legacy notification system with a unified notifier and temporary toast flow.
  - Made app bar configurable and integrated notifications actions (mark all as read, clear all).
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

grunch commented:
@BraCR10 can you address rabbit comments and fix conflicts?

BraCR10 commented:
Thanks Sir, I think that I fixed all changes suggested by rabbit, pls let me know if more changes are required

Catrya commented:
> Listo, atendere los cambios ASAP, gracias

@BraCR10 si hay algo que no te quede claro o tienes alguna duda, pregunta sin pena

BraCR10 commented:
Hola @Catrya ,  
Me surgieron algunas dudas sobre la feature de notificaciones y me gustaría comentarte lo siguiente:  

1. En la imagen que muestras aparece ese mensaje. Sin embargo, la acción `InsufficientFunds` no existe en **mostrod**.  ¿A qué se refiere esa acción?  
   - *Nota:* Esta acción la creé yo manualmente, no surgió de una acción real. Fue para probar cómo se veía la acción `CANTDO` y envié ese mensaje *hardcoded*.

2. En lugar de mostrar la pubkey de la contraparte, sería mejor mostrar el **nym** para mejorar la UX.  
   - Revisando los datos de una *Order*, solo recibo direcciones hexadecimales.  ¿Existe alguna función, utilidad o método para obtener el nym a partir de la pubkey?

3. Al comprador debe aparecerle una notificación cuando reciba este mensaje, indicándole que debe ingresar otra invoice.  
   - No me queda claro cómo implementarlo en `lib/features/order/notfiers/abstract_mostro_notifier.dart`.  ¿Podrías orientarme? ¿Cual accion seria la que lo triggea?

4. **Rating Request**  
   - ¿La notificación de *Rating Request* debe aparecerle a ambos usuarios al mismo tiempo?  
   - ¿Se dispara con cuales  acciónes `purchase-completed` o con `released`?

5. **Generación de errores**  
   - ¿Podrías guiarme un poco sobre cómo generaste estos errores relacionado con duplicaciones ?   Me ha costado reproducirlos. Aunque tengo el relay de Nostr corriendo en Docker y Mostro activo, a veces la app deja de permitir compras/ventas o directamente no envía las órdenes.  
   - Sospecho que podría ser una mala configuración.

Catrya commented:
Hola @BraCR10 disculpa la demora en responder.

> En lugar de mostrar la pubkey de la contraparte, sería mejor mostrar el nym para mejorar la UX.
        Revisando los datos de una Order, solo recibo direcciones hexadecimales. ¿Existe alguna función, utilidad o método para obtener el nym a partir de la pubkey?

Sí,  Ya existe una función para esto:
  En: `lib/shared/providers/legible_handle_provider.dart`
  Función: `deterministicHandleFromHexKey(String hexKey)`
  Provider: `nickNameProvider`
Así:   `final nym = ref.watch(nickNameProvider(pubkey));`
Esto convierte pubkeys a nyms como "shadowy-wizard", "lightning-satoshi" etc

>  Al comprador debe aparecerle una notificación cuando reciba este mensaje, indicándole que debe ingresar otra invoice.
No me queda claro cómo implementarlo en lib/features/order/notfiers/abstract_mostro_notifier.dart. ¿Podrías orientarme? ¿Cual accion seria la que lo triggea?

Cuando falla el pago de una factura primero recibe la acción `payment-failed` indicando que va a seguir intentando pagar, si no lo logra, mostrod manda un mensaje con `"action":"add-invoice"`, y dentro del payload `"status":"settled-hold-invoice"`

> ¿La notificación de Rating Request debe aparecerle a ambos usuarios al mismo tiempo?
¿Se dispara con cuales acciónes purchase-completed o con released?

No necesariamente le debe aparecer a los 2 al mismo tiempo, el vendedor la puede recibir unos segundos antes que el comprador en lo que se paga la invoice del comprador. 
Mostrod envia a los usuarios un mensaje con `"action":"rate"` a cada usuario cuando ya puede en calificar a la contraparte.

> Generación de errores
¿Podrías guiarme un poco sobre cómo generaste estos errores relacionado con duplicaciones ? Me ha costado reproducirlos. Aunque tengo el relay de Nostr corriendo en Docker y Mostro activo, a veces la app deja de permitir compras/ventas o directamente no envía las órdenes.
Sospecho que podría ser una mala configuración.

Hice algunos trades y salían duplicadas algunas notificaciones, no vi una forma de que ocurriera en todos los casos, pero sí en cada trade hubo al menos 1 notificación duplicada.

BraCR10 commented:
Gracias por tu respuesta  !!

> Generación de errores
> ¿Podrías guiarme un poco sobre cómo generaste estos errores relacionado con duplicaciones ? Me ha costado reproducirlos. Aunque tengo el relay de Nostr corriendo en Docker y Mostro activo, a veces la app deja de permitir compras/ventas o directamente no envía las órdenes.
> Sospecho que podría ser una mala configuración

A lo que me refería con esto es que en ocasiones tengo problemas para probar, porque después de un tiempo no puedo generar ordenes, cancelar, etc en local desde la UI, por eso puse lo de  `sospecho que podría ser una mala configuración ` de mi parte, hay algo que deba en consideración para probar en local aparte de levantar mostro y el relay?, como lo pruebas, desde la UI?

BraCR10 commented:
Hello @Catrya, the feature is ready for review! 

 Changes included
- Fixed duplicated notifications  
- Added role verification for certain notifications  
- Added the requested notification  
- Added responsive settings so notifications adapt to different screens
- Applied several `coderaabit` suggestions  

---

All notifications are triggered from:  
`lib/features/order/notifiers/abstract_mostro_notifier.dart`  

Please let me know if any changes are needed regarding the notification logic or if something should be adjusted.  

<img width="381" height="842" alt="image" src="https://github.com/user-attachments/assets/d255c460-f942-41b1-89ff-b7905b7c569a" />
<img width="381" height="632" alt="image" src="https://github.com/user-attachments/assets/da75bcc7-fb1f-4022-9336-1daf19092268" />

Catrya commented on lib/core/config.dart:
```diff
@@ -10,9 +10,7 @@ class Config {
-  static const String mostroPubKey =
-      '82fa8cb978b43c79b2156585bac2c011176a21d2aead6d9f7c575c005be88390';
-  //'9d9d0455a96871f2dc4289b8312429db2e925f167b37c77bf7b28014be235980';
+  static const String mostroPubKey = '9d9d0455a96871f2dc4289b8312429db2e925f167b37c77bf7b28014be235980';
```
@BraCR10 debes dejar la pubkey de mostro correcta `82fa8cb978b43c79b2156585bac2c011176a21d2aead6d9f7c575c005be88390`

BraCR10 replied:
Disculpa, ya esta arreglado.

grunch commented:
> Hi @BraCR10 after I tested I'm not getting push notifications at all, if I test on main branch I get the push notifications, also the order of showing the notifications must be inversed, the most recent on top.

anwering to myself, now I'm not receiving push notifications on main, apparently is not related to this PR but is something we must address after this PR is merge, so don't worry about that right now

BraCR10 commented:
>  the order of showing the notifications must be inversed, the most recent on top

Sorry, I did not get this point,  currently the most recent is put at the top of the notification screen, isn't it?

Catrya commented:
I think that if the hexadecimal it is the order ID, it should be displayed, because it may happen that a user has several orders at the same time and it would be better to identify which order the notification is from, but to improve the UI, please place the order ID on a line below, not to the side where it cannot be seen, and put `ID` in front of it, so that it is easily identified that it is the order ID.

grunch commented:
> Hello @grunch so I’m going to change it to show the most accurate time taking seconds into account.
> 
> The hexadecimal are the first of the order ID , however in some screen it is reduced too much due to the width so only 3 digits are seen, should I remove it?
> 

the order id is an uuid value, not a hexadecimal and yes, please remove it, when the user tap on the notification it should be redirected to that order id, but don't do that now, after merge this PR we will create a new issue to address that

grunch commented:
> Thanks for your feedback guys, so just to sum up:
> 
> 
> Is it correct ?

let's finish the logic that need to be done to merge this PR, what is missing we can address it later with new PRs, 

> 1. I should improve the timestamp of notifications using seconds, to show exactly when it happens.

yes

> 2. I have to show the order ID as another value and delete the current frame in right bottom.

just delete the id, about displaying it we can think about it later

> 3. I need to improve the title and description of the notification `Order taken` for better understanding

if you can do it quickly ok do it, but what I think is right now, don't spend much time on changing texts, we all can do those changes later, we need you finish this PR to start working in the next notification PR

BraCR10 commented:
HI @grunch , the changes have been made 

<img width="407" height="878" alt="image" src="https://github.com/user-attachments/assets/8100fd0c-dc90-40b6-9697-1488a303b835" />

grunch commented:
@BraCR10 please send a lightning invoice with amount of 300000 sats here as a comment

BraCR10 commented:
Hi @grunch ! 
Invoice:`lnbc3m1p5tvt8wpp5w7y7uqlgaev8gaxufk5n5pxazstxzmqfkgmf2pmu0m438aavtfasdqcgfexzs6jxyczqjtwwehkjcm9cqzpuxqyz5vqsp59d7306mwu4v0yy6e9hvdzfw3zanjkz6tlfz5x5t0aef39rpn30vs9qxpqysgq0g4n0k7mnmcrpsxpf5826fvawdm6yshrshqlaa7p44zd2wnuxcmpepvnk553us0vjj26uuhck44xcnq5k0vt495pha9ntzz9m0kg2eqpn20nmp`

Address: bracr10@pay.bitcoinjungle.app

grunch commented:
> lnbc3m1p5tvt8wpp5w7y7uqlgaev8gaxufk5n5pxazstxzmqfkgmf2pmu0m438aavtfasdqcgfexzs6jxyczqjtwwehkjcm9cqzpuxqyz5vqsp59d7306mwu4v0yy6e9hvdzfw3zanjkz6tlfz5x5t0aef39rpn30vs9qxpqysgq0g4n0k7mnmcrpsxpf5826fvawdm6yshrshqlaa7p44zd2wnuxcmpepvnk553us0vjj26uuhck44xcnq5k0vt495pha9ntzz9m0kg2eqpn20nmp

paid! preimage: 3da9a09f4119f38a2ad86aceff8665455620ecab656ee2bc31fef0a1086e11cf
